Make nn.Module `forward()` type annotation more permissive

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31057 Make nn.Module `forward()` type annotation more permissive**

The current signature basically will always fail to type check, because
mypy enforces that the subclass method's input types must be "wider"
than their superclass method's input types (i.e. they can vary
contravariantly). And nothing is wider than `Any`.

This change makes it so that any input params are allowed in
`forward()`. Fixes #29099

Differential Revision: [D18918034](https://our.internmc.facebook.com/intern/diff/D18918034)

suo committed:
Make nn.Module `forward()` type annotation more permissive

The current signature basically will always fail to type check, because
mypy enforces that the subclass method's input types must be "wider"
than their superclass method's input types (i.e. they can vary
contravariantly). And nothing is wider than `Any`.

This change makes it so that any input params are allowed in
`forward()`. Fixes #29099

kostmo commented:
## CircleCI build failures summary
As of commit d8c3ed4:
* **1/1** broken upstream at merge base c75bc90 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=10&sha1=c75bc9067ca7844ec3b732e843377d5b156376a8&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * Since your merge base is older than `viable/strict`, run these commands:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis (WIP)

Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=d8c3ed43e0b6809446dae0c99645718ea12fb047):

### 1 failure recognized by patterns:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3882387) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** Build ([details](https://dr.pytorch.org/build-details.html?build_id=71685121))
```
Dec 10 18:52:43 /var/lib/jenkins/workspace/third_party/fbgemm/include/fbgemm/FbgemmFP16.h:243:25: error: type qualifiers ignored on function return type [-Werror=ignored-qualifiers]
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

albanD commented on torch/nn/modules/module.pyi.in:
```diff
@@ -11,15 +11,17 @@ _grad_t = Union[Tuple[Tensor, ...], Tensor]
 # We parameter modules by the return type of its `forward` (and therefore `__call__`) method. This allows
 # type inference to infer that the return value of calling a module in the canonical way (via `__call__)` is the
-# same as the custom `forward` function of the submodule. Submodules tha wish to opt in this functionality be 
+# same as the custom `forward` function of the submodule. Submodules tha wish to opt in this functionality be
```
nit: `tha` -> `that` if you're fixing stuff on that line? :)

suo committed:
Update on "Make nn.Module `forward()` type annotation more permissive"


The current signature basically will always fail to type check, because
mypy enforces that the subclass method's input types must be "wider"
than their superclass method's input types (i.e. they can vary
contravariantly). And nothing is wider than `Any`.

This change makes it so that any input params are allowed in
`forward()`. Fixes #29099
